modules/ethereum-contract/contracts/substrate-bridge.sol

```diff
@@ -0,0 +1,327 @@
+		return (lastImportedHeader.number, lastImportedHeaderHash);
+	}
+
+	/// Returns true if header is known to the bridge.
+	/// @param headerHash Hash of the header we want to check.
+	function isKnownHeader(
+		bytes32 headerHash
+	) public view returns (bool) {
+		return headerByHash[headerHash].isKnown;
+	}
+
+	/// Returns true if finality proof is required for this header.
+	function isFinalityProofRequired(
+		bytes32 headerHash
+	) public view returns (bool) {
+		Header storage header = headerByHash[headerHash];
+		return header.isKnown
+			&& header.number > bestFinalizedHeaderNumber
+			&& header.number == header.prevSignalTargetNumber
+			&& header.validatorsSetId == bestFinalizedValidatorsSetId;
+	}
+
+	/// Import header.
+	function importHeader(
+		bytes memory rawHeader
+	) public {
+		// parse header
+		ParsedHeader memory header = parseSubstrateHeader(rawHeader);
+		require(
+			!headerByHash[header.hash].isKnown,
+			"Header is already known"
+		);
+
+		// check if we're able to coninue chain with this header
+		Header storage parentHeader = headerByHash[header.parentHash];
+		require(
+			parentHeader.isKnown && parentHeader.number == header.number - 1,
+			"Missing parent header from the storage"
+		);
+
+		// forbid appending to fork until we'll get finality proof for header that
+		// requires it
+		if (parentHeader.prevSignalTargetNumber != 0 && parentHeader.prevSignalTargetNumber == parentHeader.number) {
+			require(
+				bestFinalizedHeaderHash == header.parentHash,
+				"Missing required finality proof for parent header"
+			);
+		}
+
+		// forbid overlapping signals
+		uint256 prevSignalTargetNumber = parentHeader.prevSignalTargetNumber;
+		if (header.signal.length != 0) {
+			require(
+				prevSignalTargetNumber < header.number,
+				"Overlapping signals found"
+			);
+			prevSignalTargetNumber = header.number + header.signalDelay;
+		}
+
+		// check if parent header has emitted validators set change signal
+		uint64 validatorsSetId = parentHeader.validatorsSetId;
+		bytes32 prevSignalHeaderHash = parentHeader.prevSignalHeaderHash;
+		if (parentHeader.signal.length != 0) {
+			prevSignalHeaderHash = header.parentHash;
+			validatorsSetId = validatorsSetId + 1;
+		}
+
+		// store header in the storage
+		headerByHash[header.hash] = Header({
+			isKnown: true,
+			parentHash: header.parentHash,
+			number: header.number,
+			signal: header.signal,
+			validatorsSetId: validatorsSetId,
+			prevSignalHeaderHash: prevSignalHeaderHash,
+			prevSignalTargetNumber: prevSignalTargetNumber
+		});
+		lastImportedHeaderHash = header.hash;
+	}
+
+	/// Import finality proof.
+	function importFinalityProof(
+		uint256 finalityTargetNumber,
+		bytes32 finalityTargetHash,
+		bytes memory rawFinalityProof
+	) public {
+		// check that header that we're going to finalize is already imported
+		require(
+			headerByHash[finalityTargetHash].number == finalityTargetNumber,
```

Comment:
[SOLIDITY] I'm not against that (will consider that when preparing final version of the contract), but I see nothing criminal in re-finalizing finalized block (genesis in this case) - it just should be a noop. Do you have any other concerns to mention?

Reply:
Since `0` is used for initialized storage as well, I'd prefer to keep it on the safe side here. i.e. with `finalityTargetNumber ==0` this check would pass with arbitrarty `finalityTargetHash` (i.e. non-existent).
So nothing very specific for now, but just my internal warning side lightning up :)